src/os/LinuxPWM.hxx

```diff
@@ -0,0 +1,164 @@
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+
+class LinuxPWM : public PWM
+{
+public:
+    LinuxPWM(int chip, int channel) : chip_(chip) , channel_(channel)
+    {
+        snprintf(pwmdir,sizeof(pwmdir),
+                 "/sys/class/pwm/pwmchip%d/pwm%d",
+                 chip_,channel_);
+    }
+    void exportPin()
+    {
+        char exportname[50];
+        strcpy(exportname,pwmdir);
+        strcat(exportname,"/period");
+        if (access(exportname, F_OK) != -1)
+        {
+            set_period(1);
+            return;
+        }
+        char *p = strrchr(exportname,'/');
+        *p = '\0';
+        p = strrchr(exportname,'/');
+        p++;
+        strcpy(p,"export");
+        set_sysfs_file_value(exportname,channel_);
+        set_period(1);
+    }       
+    void set_period(uint32_t counts) override
+    {
+        char periodname[60];
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        set_sysfs_file_value(periodname,counts);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_period() override
+    {
+        char periodname[60];
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        return get_sysfs_file_value(periodname);
+    }
+    void set_duty(uint32_t counts) override
+    {
+        char duty_cyclename[60];
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        set_sysfs_file_value(duty_cyclename,counts);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_duty() override
+    {
+        char duty_cyclename[60];
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        return get_sysfs_file_value(duty_cyclename);
+    }
+    uint32_t get_period_max() override
+    {
+        return 0xffffffff;
+    }
+    uint32_t get_period_min() override
+    {
+        return 1;
+    }
+    void enable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        set_sysfs_file_value(enablename,1);
+    }
+    void disable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        set_sysfs_file_value(enablename,0);
+    }
+    bool enabled() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        return get_sysfs_file_value(enablename) == 1;
+    }
+private:
+    const uint32_t chip_;
+    const uint32_t channel_;
+    char pwmdir[40];
+    uint32_t get_sysfs_file_value(const char* basename) {
+        uint32_t value;
+        FILE *fp = fopen(basename,"r");
+        fscanf(fp,"%d",&value);
+        fclose(fp);
+        return value;
+    }
+    void set_sysfs_file_value(const char* basename, uint32_t value) {
+        FILE *fp = fopen(basename,"w");
```

Comment:
I believe @balazsracz intent was to move the strcpy/strcat into this (and method above) so that the callers would just be:
```
set_sysfs_file_value("enable", 1);
```
and internally it would be converted to pwmdir + "/" + "enable"

Reply:
see https://github.com/RobertPHeller/openmrn/pull/4
note you also need to merge openmrn master for it to work.